metricshub-engine/src/main/java/org/metricshub/engine/strategy/detection/CriterionProcessor.java

```diff
@@ -359,19 +372,116 @@ public CriterionTestResult process(@SpanAttribute("criterion.definition") SnmpGe
+			.findCriterionExtension(criterion, telemetryManager)
+			.map(ext -> {
+				CriterionTestResult r = ext.processCriterion(criterion, connectorId, telemetryManager);
+				if (r != null) {
+					r.setCriterion(criterion);
+					return r;
 				}
-				return result;
+				return CriterionTestResult.empty();
 			})
-			.orElseGet(CriterionTestResult::empty);
+			.orElse(CriterionTestResult.empty());
+
+		if (
+			recordOutputDirectory != null &&
+			!recordOutputDirectory.isBlank() &&
+			!(criterion instanceof SnmpGetCriterion) &&
+			!(criterion instanceof SnmpGetNextCriterion)
+		) {
+			persist(result, connectorId, criterion, recordOutputDirectory);
+		}
+		return result;
+		// CHECKSTYLE:ON
+	}
+
+	/**
+	 * Persists the {@link CriterionTestResult} to a YAML file.
+	 *
+	 * @param result the {@link CriterionTestResult} to persist
+	 * @param connectorId the identifier of the connector defining the criterion
+	 * @param criterion the {@link Criterion} that was processed
+	 * @param recordOutputDirectory the directory to store the results
+	 */
+	private void persist(
+		final CriterionTestResult result,
+		final String connectorId,
+		final Criterion criterion,
+		final String recordOutputDirectory
+	) {
+		// Directory where we will store the criterion results
+		final Path criterionResultOutputDirectory = Paths.get(recordOutputDirectory);
+		try {
+			Files.createDirectories(criterionResultOutputDirectory);
+
+			final Path file = criterionResultOutputDirectory.resolve(
+				telemetryManager.getHostname() + "_" + connectorId + "_" + criterion.getType() + "_criterion.yaml"
+			);
+
+			try (
+				BufferedWriter out = Files.newBufferedWriter(
+					file,
+					StandardCharsets.UTF_8,
+					StandardOpenOption.CREATE,
+					StandardOpenOption.TRUNCATE_EXISTING
+				)
+			) {
+				YAML_MAPPER.writeValue(out, result);
+				out.newLine();
+			}
+		} catch (IOException e) {
+			log.warn("Could not write CriterionTestResult to {}", criterionResultOutputDirectory, e);
+		}
+	}
+
+	/**
+	 * Reads a {@link CriterionTestResult} from the emulated criterion output directory.
+	 *
+	 * @param connectorId the identifier of the connector defining the criterion
+	 * @param criterion the {@link Criterion} to load results for
+	 * @param emulationModeCriterionOutputDirectory directory containing emulated results
+	 * @return An {@link Optional} with the {@link CriterionTestResult} if found, or empty otherwise
+	 */
+	private Optional<CriterionTestResult> readEmulatedCriterionResult(
+		String connectorId,
+		Criterion criterion,
+		String emulationModeCriterionOutputDirectory
+	) {
+		final Path outDir = Paths.get(emulationModeCriterionOutputDirectory);
+		final Path file = outDir.resolve(
+			telemetryManager.getHostname() + "_" + connectorId + "_" + criterion.getType() + "_criterion.yaml"
+		);
+
+		if (!Files.exists(file)) {
+			throw new IllegalStateException("The path " + file.getFileName() + " does not exist!");
```

Comment:
Sometimes the file is not present for some technical reason, so we shouldn't beak the engine we the IllegalStateException here.